### ISSUE TITLE
isInstanceOf is statically optimized away based on compiler knowledge about types

### ISSUE THREAD

**Issue by valenterry** [scala/bug#10379]
The `isInstaceOf` check are optimized away in the following situation:
``` scala
    class A
    trait T
    
    val casted: A with T = (new A).asInstanceOf[A with T]
    casted.isInstanceOf[A with T] // = true
```

However:
``` scala
    casted.isInstanceOf[A] // = true
    casted.isInstanceOf[T] // = false
```
(fiddle: https://scalafiddle.io/sf/rjfApD5/0)

The check is optimized away and replaced with `true`. I am not sure if that is a bug or maybe even desired (optimization) behaviour. However, when using `isInstanceOf` I would intuitively have expected a runtime-check _no matter what the compiler statically knows_. 

In any case I would expect the test for `T` and for `A with T` to both give the same result, be it true or false.

I could reproduce this in both scala 2.11.8 and 2.12.1.

**Comment by lrytz** [comment]
I don't think there's a bug here.

You cannot cast to `A with T`: a cast is a run-time type check, but there's not run-time type representing `A with T`. What happens ist that the Scala compiler erases `A with T` to `A`, so you end up with `(new A).asInstanceOf[A]`, which obviously passes. With `-Xprint:erasure`

```
scala> class A; trait T
scala> val a = new A
scala> a.asInstanceOf[A with T]
[[syntax trees at end of                   erasure]] // <console>
...
        private[this] val res1: A = $line5.$read.$iw.$iw.a().$asInstanceOf[A]();

scala> a.asInstanceOf[T]
java.lang.ClassCastException: A cannot be cast to T
```

The same is true for the `casted.isInstanceOf[A with T]` check, it gets erased to `casted.isInstanceOf[A]`.

What would be useful here is a compiler warning when a type argument of `asInstanceOf` / `isInstanceOf` cannot be represented at run time.

**Comment by sjrd** [comment]
> What would be useful here is a compiler warning when a type argument of asInstanceOf / isInstanceOf cannot be represented at run time.

Not for `asInstanceOf`! There are zillions of cases where you cast and the target type is not entirely checked, including many compiler-generated ones. Issuing warnings for such cases is not practical.

However, `isInstanceOf[A with T]` should even be a compile-time *error*, because the compiler cannot (or at least *does* not) compile it correctly, giving you a wrong result. There is no use case for this.

For example, in Scala.js, when `Foo` is a JavaScript trait, `x.asInstanceOf[Foo]` is erased, but `x.isInstanceOf[Foo]` is a compile error, instead of blindly returning `true` which is never useful.

**Comment by lrytz** [comment]
Good points, thanks @sjrd

**Comment by valenterry** [comment]
Yes, you cannot cast it at runtime. But there is a use case for it at compiletime for tagging types. See for example: https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/typeoperators.scala

So, yes, `a.asInstanceOf[A with T]` will not change the runtime value, but it will in fact change what the compiler knows (or thinks to know) about the type.

If `casted.isInstanceOf[A with T]` gets erased to `casted.isInstanceOf[A with T]` then what about `casted.isInstanceOf[T]`?

**Comment by lrytz** [comment]
> what about `casted.isInstanceOf[T]`

This cast can be checked, `T` exists at run-time, it's an interface.

**Comment by valenterry** [comment]
I see. So `casted.isInstanceOf[A with T]` is always the same as `casted.isInstanceOf[A]`. Thanks for the clarification.

**Comment by smarter** [comment]
> I see. So casted.isInstanceOf[A with T] is always the same as casted.isInstanceOf[A]. Thanks for the clarification.

No. `isInstanceOf` is much more magic than `asInstanceOf`, `isInstanceOf[A with T]` will be rewritten to `isInstanceOf[A] && isInstanceOf[T]` by the compiler:
```scala
class A
trait T

object Test {
  def foo(x: Any) = {
    x.isInstanceOf[A with T]
  }
}
```
```scala
[[syntax trees at end of                       jvm]] // isInstance.scala
package <empty> {
  class A extends Object {
    def <init>(): A = {
      A.super.<init>();
      ()
    }
  };
  abstract trait T extends Object;
  object Test extends Object {
    def foo(x: Object): Boolean = x.$isInstanceOf[A]().&&(x.$isInstanceOf[T]());
    def <init>(): Test.type = {
      Test.super.<init>();
      ()
    }
  }
}
```
From [the spec, chapter 12](http://www.scala-lang.org/files/archive/spec/2.12/12-the-scala-standard-library.html):
> The type test x.isInstanceOf[T] is equivalent to a typed pattern match

and again from [the spec, chapter 8](http://www.scala-lang.org/files/archive/spec/2.12/08-pattern-matching.html#type-patterns):
> A compound type pattern T1 with … with Tn where each Ti is a type pattern. This type pattern matches all values that are matched by each of the type patterns Ti.:


What happens in your original example is that the compiler optimises out the `isInstanceOf[A with T]` check to `true` based on its knowledge of static types, but it fails to take into account that having a static type of `A with T` does not mean that `isInstanceOf[A] && isInstanceOf[T]` will always be true due to erasure. I would consider this a bug.

**Comment by valenterry** [comment]
I will reopen and let you guys decide. ;)

**Comment by lrytz** [comment]
Thanks @smarter, nice, I wasn't aware!

To summarize the issue:

```
scala> class C; trait T
scala> val a = new C

scala> a.isInstanceOf[C with T]
res0: Boolean = false

scala> a.asInstanceOf[C with T].isInstanceOf[C with T]
res1: Boolean = true
```

It's consistent with type patterns:

```
scala> a match { case _: C with T => true; case _ => false }
res2: Boolean = false

scala> a.asInstanceOf[C with T] match { case _: C with T => true; case _ => false }
res3: Boolean = true
```

Not sure this is a bug. It requires a cast to get in such a situation, and I don't know what kind of behavior is guaranteed / allowed after casting. cc @adriaanm

**Comment by adriaanm** [comment]
Haha, cool!  Agreed that we shouldn't constant fold `a.asInstanceOf[C with T].isInstanceOf[C with T]` to `true`, but it's also true we don't guarantee much once you start casting. 

For the pro-bug case, it's the same as

```
scala> class A
defined class A

scala> class B
defined class B

scala> (new A).asInstanceOf[B].isInstanceOf[B]
java.lang.ClassCastException: A cannot be cast to B
  ... 29 elided
```

(except that, as explained, you get away with casting to `C with T`)

**Comment by lrytz** [comment]
> we shouldn't constant fold `a.asInstanceOf[C with T].isInstanceOf[C with T]` to `true`

Where would you draw the line? Which assumptions can we make about the correspondance of static and dynamic types, and which not?

**Comment by sjrd** [comment]
Well you can assume that the value is really an instance of its *erased* type, and nothing else. Pretty clear to me.

The problem is that the erasure of a type is a tad platform-dependent. In Scala.js all types that are `<: js.Any` effectively erase to `any` (or `Object` to speak a common language). Though at the same time I think we somewhat rely on the instance test being folded to true for JS type extraction to work (see tests here: https://github.com/scala-js/scala-js/blob/master/test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/TupleTest.scala).

**Comment by sjrd** [comment]
> Well you can assume that the value is really an instance of its erased type, and nothing else.

No in fact not even that is true: it can also be `null`.

**Comment by DarkDimius** [comment]
@lrytz, in Dotty we've decided to always restructure casts:
https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

`a.isInstanceOf[C & T]` is always rewritten to `a.isInstanceOf[C] && a.isInstanceOf[T]`.

Later: 
 - If `a` has a `case` flag, which indicates that it was created by patmat, we can issue a warning if one of tests is known to always evaluate to the same value: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala
 - In simplify I've tried to see how many of the null checks resulting from rewriting 
a.isInstanceOf[C] to `a != null`. I wasn't able to eliminate enough of them and this stops a lot of optimizations from triggering.

**Comment by valenterry** [comment]
> except that, as explained, you get away with casting to C with T

If I instantiate something with `new C with T` the compiler will create a new anonymous class to make that work. If I cast to `C with T` I would expect this to either fail (or at least warn) or to work by that the compiler creates a new anonymous class as he does with `new C with T`... for which I cannot see any sane use case, but it would be consistent at least.

**Comment by sjrd** [comment]
@valenterry In `new C with T`, the `C with T` is basically a class specification. In `x.isInstanceOf[C with T]`, the `C with T` part is a *type*, not a class. Those two usages need not be consistent, because they don't work on the same kind of things (class versus type). This is not limited to `with`. For example `.isInstanceOf[HashSet[_]]` is fine but `new HashSet[_]` is not, because `HashSet[_]` is a type but not a class.

**Comment by valenterry** [comment]
@sjrd I know that and I don't expect it to be consistent, but I would like to rely on that if I successfully casted something to `X`, then whatever the type of `X` is, I would expect that all of `X`s guaranteed properties (e.g. existing methods) at compile time are also working at runtime. That is, if my trait `T` would have methods, I could cast to `C with T` and call those methods, I would get runtime exceptions.

**Comment by alexeyr** [comment]
The spec also seems to say `asInstanceOf` should be consistent with pattern-matching by giving the definition:

    /** Type cast; needs to be inlined to work as given */ */
    def asInstanceOf[A]: A = this match {
      case x: A => x
      case _ => if (this eq null) this
                else throw new ClassCastException()
    }

according to which `(new A).asInstanceOf[A with T]` should fail. Perhaps the definition be should be changed along the same lines as `isInstanceOf`?

> The cast `x.asInstanceOf[T]` is equivalent to a typed pattern match
> ```
> x match {
>  case y: T′ => y
>  case _ => if (x eq null) default[T] else throw new ClassCastException()
> }
> ```
> where the type `T′` is the erasure of `T` and `default[T]` stands for the default value of `T`.

(with "erasure" and "default value" linked to the proper sections). Is this entirely correct?
